File: Python/Data Structure/Bag/linked_list_bag.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None
    
class Bag:
    def __init__(self):
        self.head = None
        self.tail = self.head
        self.len = 0

    def is_empty(self):
        return not self.head
    
    def push(self, value):
        new_node = Node(value)

        if self.is_empty():
            self.head = new_node
            self.tail = self.head
            self.len += 1
            return

        self.tail.next = new_node
        self.tail = new_node
        self.len += 1
        return
    
    def search(self, index):
        if self.is_empty() or index < 0 or index > self.len:
            return None
        
        current_node = self.head
        for i in range(1, index):
            current_node = current_node.next
        
        return current_node.value

    def log(self):
        if self.is_empty():
            return None
        current_node = self.head
        result = []

        while current_node:
            result.append(current_node.value)
            current_node = current_node.next
        
        return " ---> ".join(result)


bag = Bag()

bag.push("Saitama")
bag.push("Luffy")
bag.push("Naruto")
bag.push("Ichigo")

print(bag.log())

print(bag.search(1))
print(bag.search(2))
print(bag.search(3))
print(bag.search(4))

File: Nim/Data Structure/Bag/linked_list_bag.nim
import strutils 

type 
    Node = ref object
        value: string
        next: Node
        
type 
    BagLinkedList = ref object
        head: Node
        length: int

proc newLinkedListBag(): BagLinkedList = 
    return BagLinkedList(head: nil)

proc append(self: BagLinkedList, item: string) =
    let new_node: Node = Node(value: item, next: nil)

    if self.head.isNil:
        self.head = new_node
        self.length += 1
        return

    var current_node: Node = self.head

    while not current_node.next.isNil:
        current_node = current_node.next
    
    current_node.next = new_node
    self.length += 1

proc delete(self: BagLinkedList, index: int) = 
    if index < 0 or index > self.length + 1:
        echo "Index out of range"
        return

    if index == 1:
        let current_node: Node = self.head.next
        self.head = current_node
        self.length -= 1
        return

    var previous_node: Node = self.head

    for i in 1 ..< index-1:
        if not previous_node.isNil:
            previous_node = previous_node.next
    
    var current_node: Node = previous_node.next
    previous_node.next = current_node.next
    self.length -= 1
    return

proc search(self: BagLinkedList, index: int): string = 
    if index < 0 or index > self.length + 1:
        echo "Index out of range"
        return " "
    
    var current_node: Node = self.head
    var idx: int = 1

    while not current_node.isNil:
        if idx == index:
            return current_node.value
        current_node = current_node.next
        idx += 1

    return " "

proc get_len(self: BagLinkedList): int = 
    return self.length

proc log(self: BagLinkedList): string = 
    if self.head.isNil:
        echo "The list is empty"
        return ""

    var arr: seq[string] = @[]
    var current_node: Node = self.head

    while not current_node.isNil:
        arr.add(current_node.value)
        current_node = current_node.next
    
    return join(arr, " ---> ")

when isMainModule:
    let list: BagLinkedList = newLinkedListBag()

    let data: array[0..5, string] = ["Jack", "Mike", "Jenny", "Randall", "Natasha", "Milley"]

    for i in 0..<data.len:
        list.append(data[i])
    
    echo list.log()

    list.delete(5)

    echo list.log()

    echo list.search(data.len-3);
    

